lib/mllp/sender.ex

```diff
@@ -1,32 +1,24 @@
 defmodule MLLP.SenderContract do
   @type error_type :: :connect_failure | :send_error | :recv_error
-  @type error_reason :: :closed | :timeout | :no_socket | :inet.posix()
+  @type error_reason :: :closed | :timeout | :not_connected | :inet.posix()
 
   @type sender_error :: %{type: error_type(), reason: error_reason()}
 
-  @callback send_hl7_and_receive_ack(
+  @type send_options :: %{
+          optional(:expect_reply) => boolean(),
+          optional(:timeout) => non_neg_integer(),
+          optional(:wrap_message) => boolean(),
+          optional(:reply_timeout) => non_neg_integer() | :infinity
+        }
+
+  @callback send(
               pid :: pid,
               payload :: HL7.Message.t(),
-              timeout :: non_neg_integer()
-            ) ::
-              MLLP.Ack.ack_verification_result() | {:error, sender_error()}
-  @callback send_hl7(pid :: pid, payload :: HL7.Message.t()) ::
-              {:ok, :sent} | {:error, sender_error()}
-  @callback send_non_hl7_and_receive_reply(
-              pid :: pid,
-              payload :: String.t(),
-              timeout :: non_neg_integer()
+              options :: send_options()
             ) ::
-              {:ok, String.t()} | {:error, sender_error()}
-  @callback send_non_hl7(pid :: pid, payload :: String.t()) ::
-              {:ok, :sent} | {:error, sender_error()}
-  @callback send_raw_and_receive_reply(
-              pid :: pid,
-              payload :: String.t(),
-              timeout :: non_neg_integer()
-            ) ::
-              {:ok, String.t()} | {:error, sender_error()}
-  @callback send_raw(pid :: pid, payload :: String.t()) :: {:ok, :sent} | {:error, sender_error()}
+              {:ok, binary() | String.t()}
```

Comment:
wouldn't we always return ACK message or error? In what cases we return `{:ok, binary()}`

Reply:
Good catch.  That's actually redundant. It should be just `{:ok, String.t()}`  `String.t()` is equivalent to `binary()`

Reply:
> wouldn't we always return ACK message or error? 

I might have misunderstood. We would return a binary (or String.t())` if verify_reply is false (i.e., send raw and receive raw reply).

Reply:
I take `raw` message here still an HL7 message but as `binary()` rather than type HL7.Message.t(). I think `send` should accept both `binary` and `HL7.Message.t()`. If message received is an `HL7.Message.t()`, then we should always return verified ACK. If we receive `binary()` raw hl7, we return `binary()` response. I think, we don't need `verify_reply` option